### PR TITLE
UPI E2E Test

### PR DESCRIPTION
**Pull request by pradithya**
Add end to end tests for traffic splitting and standard ensembler when using UPI protocol.  The PR also update fiber version to include https://github.com/gojek/fiber/pull/7 which minimize the number of unmarshalling in Turing router's components. 

## E2E Test Setup

The end to end test setup reuse existing infrastructure with the difference of adding 2 separate mock services for UPI: `mockserver-upi-a` and `mockserver-upi-control` . Two separates service is required as the service endpoint can't be changed unlike the HTTP mock server. 

## Other Changes
- Reduce most router cpu requests in e2e test to `100m` (from `200m`) and increases ensembler and enricher cpu request to `200m` (from `100m`). The e2e test is more stable with this resource request.

### CODE REVIEW

**Comment by krithika369** [comment]
> It's likely that the e2e test consume more cpu due to additional mock upi server being deployed.

I'm a bit doubtful of this change. Recently, @terryyylim mentioned that there was a scenario where he had to increase the CPU allocation for the resources created by the tests because it was too low, otherwise the pods would not start up. Is the new `20m` too low? Should we try halving the previous values first (`100m`)?

Also,

> I encountered more flaky e2e test without this changes.

Does it mean the tests are still flaky? 😕

File: api/e2e/test/12_deploy_upi_router_with_std_ensembler_test.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+				globalTestContext.KServiceDomain,
+			)
+			assert.Equal(t, expectedEndpoint, router.Endpoint)
+
+			conn, err := grpc.Dial(router.Endpoint, grpc.WithTransportCredentials(insecure.NewCredentials()))
+			assert.NoError(t, err)
+			defer conn.Close()
+
+			client := upiv1.NewUniversalPredictionServiceClient(conn)
+			t.Log("Testing router endpoint with request that satisfy traffic rule rule-1")
```
Comment (leonlnj) [inline api/e2e/test/12_deploy_upi_router_with_std_ensembler_test.go:52]: shouldn't be traffic rule 😄

File: api/e2e/test/helpers_api_test.go [hunk 2]
```diff
@@ -177,6 +177,10 @@ func withDeployedRouter(
 
 	responsePayload, err := io.ReadAll(resp.Body)
 	require.NoError(t, err)
+	t.Logf("Response %s", responsePayload)
+
+	require.True(t, resp.StatusCode >= http.StatusOK && resp.StatusCode <= http.StatusPartialContent,
```
Comment (leonlnj) [inline api/e2e/test/helpers_api_test.go:182]: are there cases where it return 2XX > 200?

File: infra/cluster-init/init.sh [hunk 3]
```diff
@@ -47,14 +47,14 @@ function install_knative {
     kubectl apply \
         -f "https://github.com/knative/serving/releases/download/knative-v${KNATIVE_VERSION:-1.0.1}/serving-hpa.yaml"
 
-    kubectl apply \
-        -f knative-configmaps/config-features.yaml
-
-    local core_apps=("activator" "autoscaler" "controller" "webhook")
+    local core_apps=("activator" "autoscaler" "controller" "webhook" "domain-mapping" "domainmapping-webhook")
```
Comment (leonlnj) [inline infra/cluster-init/init.sh:50]: just for my understanding what this addition for? `"domain-mapping" "domainmapping-webhook"`

Comment (pradithya) [reply infra/cluster-init/init.sh:50]: This is additional component that knative introduces recently. Adding these 2 for completeness to ensure that knative is completely ready.

File: api/e2e/test/helpers_api_test.go (hunk 2 above)
Comment (pradithya) [reply api/e2e/test/helpers_api_test.go:182]: It's a bit defensive coding, not really sure whether there is such case for Turing.

File: api/e2e/test/11_deploy_upi_router_with_traffic_rules_test.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+				globalTestContext.KServiceDomain,
+			)
+			assert.Equal(t, expectedEndpoint, router.Endpoint)
+
+			conn, err := grpc.Dial(router.Endpoint, grpc.WithTransportCredentials(insecure.NewCredentials()))
+			assert.NoError(t, err)
+			defer conn.Close()
+
+			client := upiv1.NewUniversalPredictionServiceClient(conn)
+			t.Log("Testing router endpoint with request that satisfy traffic rule rule-1")
```
[the remark this answers is not in the document]
Comment (pradithya) [reply api/e2e/test/11_deploy_upi_router_with_traffic_rules_test.go:45]: Yes correct!

File: api/e2e/test/helpers_api_test.go (hunk 2 above)
Comment (leonlnj) [reply api/e2e/test/helpers_api_test.go:182]: ok, was just curious if there anything that happen that led to this

File: api/e2e/test/12_deploy_upi_router_with_std_ensembler_test.go (hunk 1 above)
Comment (pradithya) [reply api/e2e/test/12_deploy_upi_router_with_std_ensembler_test.go]: Updated

**Comment by terryyylim** [comment]
> > Recently, @terryyylim mentioned that there was a scenario where he had to increase the CPU allocation for the resources created by the tests because it was too low
> 
> Which test is it? I encountered [another flaky](https://github.com/caraml-dev/turing/actions/runs/3272521270/jobs/5383668469) test caused by `no such host` error.
> 
> ```
>     01_create_router_test.go:54: 
>         	Error Trace:	/home/runner/work/turing/turing/api/e2e/test/01_create_router_test.go:54
>         	            				/home/runner/work/turing/turing/api/e2e/test/helpers_test.go:71
>         	            				/home/runner/work/turing/turing/api/e2e/test/01_create_router_test.go:45
>         	            				/home/runner/work/turing/turing/api/e2e/test/helpers_api_test.go:292
>         	            				/home/runner/work/turing/turing/api/e2e/test/01_create_router_test.go:35
>         	Error:      	Not equal: 
>         	            	expected: 200
>         	            	actual  : 500
>         	Test:       	TestEndToEnd/EndToEnd/CreateRouter_KnativeServices
>         	Messages:   	Unexpected response (code 500): Post "http://e2e-experiment-3272521270-turing-enricher-1.default.svc.cluster.local/anything": dial tcp: lookup e2e-experiment-3272521270-turing-enricher-1.default.svc.cluster.local: no such host
> ```
> 
> In this case I think the problem is with either knative or istio cpu request.

There were no particular tests that failed because of this. It's more that the enricher and ensembler's `queue-proxy` containers were not starting up, resulting in tests not being able to start running. This was resolved when I increased resource_request configuration for each router.

**Comment by pradithya** [comment]
> There were no particular tests that failed because of this. It's more that the enricher and ensembler's queue-proxy containers were not starting up, resulting in tests not being able to start running. This was resolved when I increased resource_request configuration for each router.

Yeah it looks like it's happening for enricher and ensembler only. I try increasing it's cpu request from 100m -> 200m . What is your recommended value @terryyylim ?

**Comment by pradithya** [comment]
> This was resolved when I increased resource_request configuration for each router.
Looks like increasing cpu request for ensembler and enricher is enough. The test is more stable (at least 3 times without failing)

**Comment by terryyylim** [comment]
> > There were no particular tests that failed because of this. It's more that the enricher and ensembler's queue-proxy containers were not starting up, resulting in tests not being able to start running. This was resolved when I increased resource_request configuration for each router.
> 
> Yeah it looks like it's happening for enricher and ensembler only. I try increasing cpu request from 100m -> 200m . What is your recommended value @terryyylim ?

@pradithya I increased it to 200m too! 👍🏻

**Comment by krithika369** [comment]
> @terryyylim @krithika369 @leonlnj Good to merge?

I didn't review the PR fully yet, I'll do it this morning.

File: api/e2e/test/12_deploy_upi_router_with_std_ensembler_test.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+				globalTestContext.KServiceDomain,
+			)
+			assert.Equal(t, expectedEndpoint, router.Endpoint)
+
+			conn, err := grpc.Dial(router.Endpoint, grpc.WithTransportCredentials(insecure.NewCredentials()))
+			assert.NoError(t, err)
+			defer conn.Close()
+
+			client := upiv1.NewUniversalPredictionServiceClient(conn)
+			t.Log("Testing router endpoint with request that satisfy traffic rule treatment-1")
```
Comment (krithika369) [inline api/e2e/test/12_deploy_upi_router_with_std_ensembler_test.go:42]: Nit: `that satisfy traffic rule treatment-1`-> `that generates treatment-a` ?

**Review by leonlnj — APPROVED** [review]
LGTM

File: sdk/turing/generated/model/traffic_rule_condition.py [hunk 6]
```diff
@@ -115,7 +115,7 @@ def __init__(self, field_source, field, values, *args, **kwargs):  # noqa: E501
 
         Args:
             field_source (FieldSource):
-            field (str): If `field_source` is `header`, then `field` should contain the name of the request header, otherwise, if `field_source` is `payload`, then `field` should be a valid json path 
+            field (str): For HTTP_JSON protocol, the valid `field_source` are `header` and `payload`. Whereas, for UPIV1 protocol the valid `field_source` are `header` and `prediction_context` If `field_source` is `header`, then `field` should contain the name of the request header, otherwise, if `field_source` is `payload`, then `field` should be a valid json path. If `field_source` is `header`, then `field` should contain the name of the request header, otherwise, if `field_source` is `prediction_context`, then `field` should contains variable name.
```
Comment (krithika369) [inline sdk/turing/generated/model/traffic_rule_condition.py:118]: > If `field_source` is `header`, then `field` should contain the name of the request header

I think this part of the comment is repeated. From the way the description is written [here](https://github.com/caraml-dev/turing/blob/grpc_upi/api/api/specs/routers.yaml#L723), it's not clear that the lines 2 and 3 correspond to different protocols and I'm not sure it's important to distinguish. Shall we simply remove the duplicated line?

>           description: >
            For HTTP_JSON protocol, the valid `field_source` are `header` and `payload`. Whereas, for UPIV1 protocol the valid `field_source` are `header` and `prediction_context`
            If `field_source` is `header`, then `field` should contain the name of the request header. If `field_source` is `payload`, then `field` should be a valid json path. If `field_source` is `prediction_context`, then `field` should contains variable name.

**Review by krithika369 — APPROVED** [review]
LGTM! Thanks for digging into the flaky test cases too.

File: api/e2e/test/12_deploy_upi_router_with_std_ensembler_test.go (hunk 5 above)
Comment (pradithya) [reply api/e2e/test/12_deploy_upi_router_with_std_ensembler_test.go]: Updated

File: sdk/turing/generated/model/traffic_rule_condition.py (hunk 6 above)
Comment (pradithya) [reply sdk/turing/generated/model/traffic_rule_condition.py]: Yeah, rephrased the documentation.

**Pull request merged**